src/transformers/modeling_utils.py

```diff
@@ -2093,7 +2094,35 @@ def save_pretrained(
 
         # Save the model
         if state_dict is None:
-            state_dict = model_to_save.state_dict()
+            # if model parameters are offloaded, onload and send state dicts to CPU
+            if hasattr(model_to_save, "_hf_hook") and isinstance(model_to_save._hf_hook, AlignDevicesHook):
+                state_dict = {}
+                for name, module in model_to_save.named_modules():
+                    if name == "":
+                        continue
+                    if hasattr(module, "_hf_hook") and isinstance(module._hf_hook, AlignDevicesHook):
+                        # onload meta tensors to execution device
+                        try:
+                            module._hf_hook.pre_forward(module)
```

Comment:
What if instead we modify `module._hf_hook.device` to be `cpu` directly instead here, so it won't be placed on CUDA. Still offloads, but skips a step and saves memory. 

Then after `post_forward`, just replace that value back.

Reply:
Good idea! That simplifies things somewhat, and I have implemented this by assigning and re-assigning the `module._hf_hook.execution_device`